src/sources/mongodb_metrics/mod.rs

```diff
@@ -0,0 +1,1065 @@
+                tags!(self.tags, "type" => r#type),
+            ));
+        }
+
+        // op_counters_total
+        for (r#type, value) in status.opcounters_repl {
+            metrics.push(self.create_metric(
+                "op_counters_total",
+                counter!(value),
+                tags!(self.tags, "type" => r#type),
+            ));
+        }
+
+        Ok(metrics)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn generate_config() {
+        crate::test_util::test_generate_config::<MongoDBMetricsConfig>();
+    }
+}
+
+#[cfg(all(test, feature = "mongodb_metrics-integration-tests"))]
+mod integration_tests {
+    use super::*;
+    use crate::{test_util::trace_init, Pipeline};
+    use futures::compat::Stream01CompatExt;
+    use tokio::time::{timeout, Duration};
+
+    async fn test_instance(endpoint: &'static str) {
+        let host = "localhost";
+        let namespace = "vector_mongodb";
+
+        let (sender, recv) = Pipeline::new_test();
+        let mut recv = recv.compat();
+
+        tokio::spawn(async move {
+            MongoDBMetricsConfig {
+                endpoint: endpoint.to_owned(),
+                scrape_interval_secs: 15,
+                namespace: namespace.to_owned(),
+            }
+            .build(
+                "default",
+                &GlobalOptions::default(),
+                ShutdownSignal::noop(),
+                sender,
+            )
+            .await
+            .unwrap()
+            .compat()
+            .await
+            .unwrap()
+        });
+
+        let event = timeout(Duration::from_secs(3), recv.next())
+            .await
+            .expect("fetch metrics timeout")
+            .expect("failed to get metrics from a stream");
+        let mut events = vec![event];
+        loop {
+            match timeout(Duration::from_millis(10), recv.next()).await {
+                Ok(Some(event)) => events.push(event),
+                Ok(None) => break,
+                Err(_) => break,
+            }
+        }
+
+        assert!(events.len() > 100);
+        for event in events {
+            let metric = event.expect("Valid Event").into_metric();
+            // validate namespace
+            assert!(metric.name.starts_with(&format!("{}_", namespace)));
+            // validate timestamp
+            let timestamp = metric.timestamp.expect("existed timestamp");
+            assert!((timestamp - Utc::now()).num_seconds() < 1);
+            // validate basic tags
+            let tags = metric.tags.expect("existed tags");
+            assert_eq!(tags.get("endpoint").map(String::as_ref), Some(endpoint));
+            assert_eq!(tags.get("host").map(String::as_ref), Some(host));
+        }
+    }
+
+    #[tokio::test]
+    async fn fetch_metrics_mongod() {
+        trace_init();
+        test_instance("mongodb://localhost:27017").await;
+    }
+
+    // TODO
+    // #[tokio::test]
+    // async fn fetch_metrics_mongos() {
```

Comment:
Do we need a follow-up issue for supporting this?

Reply:
Yeah, I'll put a line to the first message about the test for `mongos`.